work_queue/src/batch_job_sandbox.c

```diff
@@ -0,0 +1,210 @@
+#include "batch_job.h"
+#include "batch_job_internal.h"
+#include "debug.h"
+#include "process.h"
+#include "macros.h"
+#include "stringtools.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <signal.h>
+#include <ctype.h>
+
+#define CHAR_BUFF_LEN 4096
+
+static struct nvpair *output_files_list;
+static struct nvpair *task_sandbox_name_list;
+
+static batch_job_id_t batch_job_sandbox_submit (struct batch_queue *q, const char *cmd, const char *extra_input_files, const char *extra_output_files, struct nvpair *envlist )
+{
+	batch_job_id_t jobid;
+
+    // create task sandbox
+	const char *local_task_dir = nvpair_lookup_string(envlist, "local_task_dir");
+	char *template = string_format("%s/task.sandbox.XXXXXX", local_task_dir);
+    char *task_sandbox_name = mkdtemp(template);
+
+    if (output_files_list == NULL)
+    	output_files_list = nvpair_create();
+    if (task_sandbox_name_list == NULL) 
+        task_sandbox_name_list = nvpair_create();
+
+	fflush(NULL);
+	jobid = fork();
+
+	if(jobid > 0) {
+
+		debug(D_BATCH, "started process %" PRIbjid ": %s", jobid, cmd);
+		struct batch_job_info *info = malloc(sizeof(*info));
+		memset(info, 0, sizeof(*info));
+		info->submitted = time(0);
+		info->started = time(0);
+		itable_insert(q->job_table, jobid, info);
+
+        // store the extra output files and task sandbox name
+        char *jobid_str = string_format("%d", (int)jobid);
+        nvpair_insert_string(output_files_list, jobid_str, extra_output_files);
+        nvpair_insert_string(task_sandbox_name_list, jobid_str, task_sandbox_name);
+
+        while(isspace(*extra_input_files)) extra_input_files++;
+        char *p;
+        char *input_files = strdup(extra_input_files); 
+        p = strtok(input_files, ","); 
+        while (p != NULL) {
+
+            char *orig_file; 
+            char *symlink_path;
+            
+            // move input files into task sandbox
+            if (p[0] != '/') {
+                char cwd[CHAR_BUFF_LEN];
+            	getcwd(cwd, sizeof(cwd));
```

Comment:
getcwd is an expensive operation, do it once outside of the loop.
or evewn better, can you make it work without getcwd?